ui/fancy/src/js/ui.js

```diff
@@ -0,0 +1,175 @@
+    document.body.appendChild(gridElement);
+    for (let y = 0; y < this.game.height; ++y) {
+      const rowElement = document.createElement('div');
+
+      rowElement.classList.add('row');
+      gridElement.appendChild(rowElement);
+      for (let x = 0; x < this.game.width; ++x) {
+        const blockElement = document.createElement('div');
+
+        blockElement.classList.add('block');
+        blockElement.addEventListener('click', (ev) => {
+          ev.preventDefault();
+          this.action('swap', { x, y });
+        });
+        rowElement.appendChild(blockElement);
+        this.blocks.push(blockElement);
+      }
+    }
+
+    this.blocks[0].classList.add('swapper');
+    this.blocks[1].classList.add('swapper');
+
+    const previewRowElement = document.createElement('div');
+
+    previewRowElement.classList.add('row', 'preview');
+    gridElement.appendChild(previewRowElement);
+    for (let x = 0; x < this.game.width; ++x) {
+      const blockElement = document.createElement('div');
+
+      blockElement.classList.add('block');
+      blockElement.addEventListener('click', (ev) => {
+        ev.preventDefault();
+        this.action('addRow');
+      });
+      previewRowElement.appendChild(blockElement);
+      this.previewBlocks.push(blockElement);
+    }
+  }
+
+  installEventListeners() {
+    window.addEventListener('keydown', (ev) => {
+      const actionName = keyBindingMap[ev.key];
+
+      if (typeof actionName === 'string') {
+        ev.preventDefault();
+        this.action(actionName);
+      }
+    });
+    window.addEventListener('gamepadconnected', (ev) => {
+      this.gamepadSupport.devices[ev.gamepad.index] = ev.gamepad;
+      if (Object.keys(this.gamepadSupport.devices).length === 1) {
+        this.installGamepadListener();
+      }
+    });
+    window.addEventListener('gamepaddisconnected', (ev) => {
+      delete this.gamepadSupport.devices[ev.gamepad.index];
+    });
+  }
+
+  installGameLoop() {
+    this.update(this.game.step());
+    this.gameLoop = window.setInterval(() => {
+      if (this.isGameRunning) {
+        this.update(this.game.step());
+      } else {
+        window.clearInterval(this.gameLoop);
+        this.gameLoop = null;
+      }
+    }, 1000 / this.frameRate);
+  }
+
+  installGamepadListener() {
+    const gamepadListener = () => {
+      for (let index in this.gamepadSupport.devices) {
+        const device = this.gamepadSupport.devices[index];
+        const timestamp = device.timestamp;
+        const previousTimestamp = this.gamepadSupport.previousTimestamps[index];
+
+        if (previousTimestamp && previousTimestamp === timestamp) {
+          continue;
+        }
+        this.gamepadSupport.previousTimestamps[index] = timestamp;
+
+        for (let i = 0; i < device.buttons.length; ++i) {
+          if (!device.buttons[i].pressed) {
+            continue;
+          } else if (i === 0) {
+            this.action('swap');
+          } else if (i === 1) {
```

Comment:
Index of shoulder buttons varies between different game pads so there is no portable way to do so unless some third party library containing database of different game pads is included in the user interface.